draft-ietf-anima-brski-prm.md

```diff
@@ -1238,48 +1268,60 @@ In {{!RFC8995}}, the pledge generates PVR as CMS-signed JSON and PER as PKCS#10
 
-After receiving the PVR from Registrar-Agent, the registrar SHALL perform the verification as defined in {{Section 5.3 of RFC8995}}.
-In addition, the registrar SHALL verify the following parameters from the PVR:
+TODO: Needs MASA modifications for JWS Voucher. It could work with existing MASAs if RVR were CMS.
 
-* agent-provided-proximity-registrar-cert: MUST contain registrar's own registrar LDevID certificate to ensure the registrar in proximity of the Registrar-Agent is the desired registrar for this PVR.
 
-* agent-signed-data: The registrar MUST verify that the Registrar-Agent provided data has been signed with the private key corresponding to the EE (RegAgt) certificate indicated in the "kid" JOSE header parameter.
-  The registrar MUST verify that the LDevID(ReAgt) certificate, corresponding to the signature, is still valid.
-  If the certificate is already expired, the registrar SHALL reject the request.
-  Validity of used signing certificates at the time of signing the agent-signed-data is necessary to avoid that a rogue Registrar-Agent generates agent-signed-data objects to onboard arbitrary pledges at a later point in time, see also {{sec_cons_reg-agt}}.
-  The registrar MUST fetch the EE (RegAgt) certificate, based on the provided SubjectKeyIdentifier (SKID) contained in the "kid" header parameter of the agent-signed-data, and perform this verification.
-  This requires, that the registrar has access to the EE (RegAgt) certificate data (including intermediate CA certificates if existent) based on the SKID.
-  Note, the registrar may have stored the EE (RegAgt) certificate if used during TLS establishment between Registrar-Agent and registrar or it may be provided via a repository.
+After receiving the voucher from the MASA, the registrar SHOULD evaluate it for transparency and logging purposes as outlined in {{Section 5.6 of !RFC8995}}.
+The registrar then prepares the Voucher artifact to be provided via the registrar-agent to the pledge by converting to Registrar-Countersigned Voucher (Voucher') as described in {{voucher-artifact}}.
 
-If the registrar is unable to validate the PVR, it SHOULD respond with a HTTP 4xx/5xx error code to the Registrar-Agent.
+After a successful backend interaction, the registrar MUST reply with the Registrar-Countersigned Voucher artifact (Voucher') as defined in {{voucher-artifact}} in the body of a 200 OK response.
+In the response header, the Content-Type field MUST be set to `application/voucher-jws+json` as defined in {{!I-D.ietf-anima-jws-voucher}}.
 
-The following 4xx client error codes SHOULD be used:
+If the domain registrar is unable to return the Voucher, it SHOULD respond with an HTTP server error code.
 
-* 403 Forbidden: if the registrar detected that one or more security related parameters are not valid or if the pledge-provided information could not be used with automated allowance.
+TODO: proposals? 503 temp not avail to repeat?
 
```

Comment:
yes, 503 sounds good for non-availability. But the 4xx error codes are still necessary for the other error cases to my understanding.

Reply:
Yes, they are a bit further up when the request from the client is evaluated; only there 4xx client error codes make sense.
This TODO is where the agent has to send back a response to a valid request; hence, only 5xx server error codes make sense here.

Reply:
okay, understood. I need to read the final text and the diff in parallel to not miss the parts above ;-)

Reply:
Have a look at the new list. I also declared it with "SHOULD", so stronger than the "MAY" for the client errors toward the pledge, as the client errors toward RegAgt also had "SHOULD" and it made sense to me, as we have more control over the RegAgt--Registrar communication than over pledges.

Or should the MAY become SHOULD?

Reply:
I reviewed the list of error codes. They are fine from my point of view. I also support the SHOULD statement there.

Reply:
Good, changing to SHOULD.

Related: Since the RegAgt is triggering, what is expected to happen when a tPER arrives before any tPVR was seen by the pledge? There is only a weak dependency on the created-on timestamp, but for clients with RTC, it could work.
If tPER must be after an initial tPVR, we could introduce a 409 Conflict response.

Reply:
We haven't actually thought about this type of race condition as we expected the registrar-agent would wait until the PVR has been received. At least this was the intention for the flow chart. 

Also, if re-enrollment would be done, the pledge would also not see the tPVR before  the tPER.

Do we need to add a statement that described the sequential approach?

Reply:
There is no explicit normative statement and no technical restriction that the tPER must be after the tPVR. And there is a statement using "simultaneously" in the context of bulk bootstrapping. Hence, I would add a clear statement that while the order of exchanges differs, it still must follow the sequence described in Section 7.

> The Registrar-Agent MUST begin the sequence of exchanges by triggering the pledge for the Pledge Voucher-Request (PVR).
This assumes that the Registrar-Agent has already discovered ...

Reply:
That addresses the point. The simultaneous was intended to explain that the registrar-agent may onboard several pledges at the same time, always keeping the sequence of messages. But this makes it clear.

Reply:
Done